### DISCUSSION TITLE
Injecting components  - TypeError: {}.resolve is not a function

### DISCUSSION THREAD

**Discussion by khanakia** [mdx-js/mdx#2407] [Q&A]
I am trying to inject the Provider component in Next.js 14 using this example https://mdxjs.com/guides/injecting-components/

Also I have tried removing the `import.meta.resolve` keyword it still does not works.

```tsx
async function getContent() {
  const code = String(await compile(`
  import Tooltip from '@/raw/components/Tooltip';

  Hello

<Tooltip text='this is a sentence'>
  Hover over this text and see for yourself.
</Tooltip>
  `, {
    outputFormat: 'function-body',
    development: false,
    providerImportSource: import.meta.resolve('../../../../mdx-components.js')
    /* …otherOptions */
  }))

  const mdxModule = await run(code, {...runtime, baseUrl: import.meta.url})
  return {
    MDXContent: mdxModule.default,
  }
  
}
```

```
 ⨯ src/app/docs/tooltip/page.tsx (26:38) @ resolve
 ⨯ TypeError: {}.resolve is not a function
    at getContent (./src/app/docs/tooltip/page.tsx:33:36)
    at PrivacyNextPage (./src/app/docs/tooltip/page.tsx:46:34)
    at async Promise.all (index 0)
    at async Promise.all (index 0)
    at async Promise.all (index 0)
digest: "1493623488"
  24 |     outputFormat: 'function-body',
  25 |     development: false,
> 26 |     providerImportSource: import.meta.resolve('../../../../mdx-components.js')
     |                                      ^
  27 |     /* …otherOptions */
  28 |   }))
  29 |

```

**Comment by wooorm** [comment]
a) you likely don’t need to follow that guide. That guide is for when you want to do yourself what Next already does for you. Use next/mdx.
b) Next doesn’t support `import.meta.resolve`. Again, you don‘t need to do this. But if you’d have to do this, either make sure Next supports `import.meta.resolve` or use an alternative such as [`import-meta-resolve`](https://github.com/wooorm/import-meta-resolve/blob/main/index.js), or pass a file URL (`new URL(...).href`)

**Reply by khanakia** [reply]
Yes next.js does it automatically only when we import `import PrivacyPage from './Privacy.mdx';` pages like this and render them

But I have files sitting somewhere and I am using components inside mdx files which are not resolving automatically only if you import directly then Next.js resolves the components inside the MDX file automatically.

So I have no idea how to i render mdx file with components inside by parsing them manually as `<MDXContent />`

**Reply by wooorm** [reply]
> But I have files sitting somewhere and I am using components inside mdx files which are not resolving automatically

Then show how you do this. I don’t know what you are doing.
Where are your MDX files? What is that `@`? What is in `mdx-components`? Where do you use `getContent`?
And I posted a `b)` above? Use `import-meta-resolve`. Use `new URL`

**Reply by khanakia** [reply]
1. I installed the `import-meta-resolved` and used it but got another error now `TypeError: _provideComponents is not a function`

I have created a very minimalist repo with just 2 files https://github.com/khanakia/nextjs_mdx_error/


1.  `/src/app/hello/page.mdx` resolves to `http://localhost:3000/hello` and correctly renders the component with Button as Next.js is doing its magic and resolving the Button component automatically

![Screenshot 2023-11-27 at 6 10 43 PM](https://github.com/mdx-js/mdx/assets/2174170/444cd492-2afa-4e0d-8c5b-d09858bc5826)

2. Now I want to achieve the same functionality as next.js but this time I want to manually parse the MD Code you can check the code here `/src/app/page.tsx` points to http://localhost:3000

![Screenshot 2023-11-27 at 6 12 10 PM](https://github.com/mdx-js/mdx/assets/2174170/808dd6f2-153a-4919-8612-1d52b5b30f6c)


## Conclusion

I want to achieve the same functionality as Next.js and automatically resolve the imported components.

**Reply by wooorm** [reply]
```diff
@@ -1,28 +1,27 @@
 import {Fragment, useEffect, useState} from 'react'
 import {compile, run} from '@mdx-js/mdx'
 import * as runtime from 'react/jsx-runtime'
-import {resolve} from 'import-meta-resolve'
+import {useMDXComponents} from '../../mdx-components.ts'
 
 async function getContent() {
   const code = String(await compile(`
   import Button from '@/components/button';
 
   Hello World !
   
   <Button />
   `, {
     outputFormat: 'function-body',
     development: false,
-    // providerImportSource: import.meta.resolve('../../../../mdx-components.ts')
-    providerImportSource: resolve('../../../../mdx-components.ts', import.meta.url)
+    providerImportSource: '#'
   }))
 
-  const mdxModule = await run(code, {...runtime, baseUrl: import.meta.url, Fragment: Fragment})
+  const mdxModule = await run(code, {...runtime, useMDXComponents, baseUrl: import.meta.url, Fragment: Fragment})
 
   return {
     MDXContent: mdxModule.default,
   }
   
 }
 
 export default async function PrivacyNextPage() {
```

**Reply by khanakia** [reply]
The resolve error is gone 😄  and now it showing this `Error: Expected component `Button` to be defined: you likely forgot to import, pass, or provide it.`

![Screenshot 2023-11-27 at 6 56 17 PM](https://github.com/mdx-js/mdx/assets/2174170/69c7663f-6b6d-43bf-af81-ddcfefdafb69)


2. Also there is one more typesafe error 
![Screenshot 2023-11-27 at 6 55 52 PM](https://github.com/mdx-js/mdx/assets/2174170/c07ce3b1-943f-4be6-9a58-5b498cd879c9)

**Reply by wooorm** [reply]
But like, you don‘t need a provider?
Where do you use `MDXContent`?
Pass components there.

**Reply by wooorm** [reply]
> and now it showing this Error: Expected component Button to be defined: you likely forgot to import, pass, or provide it.

You haven’t defined `Button`. Where should it magically come from? Pass it, as the error says.

**Reply by wooorm** [reply]
> Also there is one more typesafe error

Yes, you have a type error because your types are not correct. Type your `useMDXComponents` correctly

**Reply by khanakia** [reply]
Yes, I see your point i can pass the Button manually but can I resolve it automatically somehow just the same way Next.js does?

**Reply by khanakia** [reply]
The types are correct  actually the `run` function needs the `MDXComponents` itself but you asked me to pass the `useMDXComponents(components: MDXComponents)` which accepts the `components param` that's why types are mismatched

**Reply by wooorm** [reply]
> i can pass the Button manually but can I resolve it automatically somehow just the same way Next.js does?

Why not put it in `mdx-components`? Why else do you want that file?

**Reply by wooorm** [reply]
> the run function needs the MDXComponents

No. `run` does not need `MDXComponents`.
Please look at your own function and our types. They are different.
https://github.com/mdx-js/mdx/blob/efeb4791ad89da9c23b44d4c79ed4c78b003a03e/packages/mdx/lib/util/resolve-evaluate-options.js#L43

**Comment by khanakia** [comment]
I modified the below code and working fine.

But the whole reason I wanted to do the provider or injection approach so I do not have to Pass the components manually.
I just want the button automatically injected just the way Next.js does e.g. `src/app/hello/page.mdx`

```tsx
import { Fragment } from 'react';
import { compile, run } from '@mdx-js/mdx';
import * as runtime from 'react/jsx-runtime';
import Button from '@/components/button';

async function getContent() {
  const code = String(
    await compile(
      `
        Hello World !
        <Button />
        `,
      {
        outputFormat: 'function-body',
        development: false,
      },
    ),
  );

  const mdxModule = await run(code, {
    ...runtime,
    baseUrl: import.meta.url,
    Fragment: Fragment,
  });

  return {
    MDXContent: mdxModule.default,
  };
}

export default async function PrivacyNextPage() {
  const { MDXContent } = await getContent();
  return (
    <article className='prose mx-auto'>
      <MDXContent
        components={{
          Button,
        }}
      />
    </article>
  );
}
```

**Reply by wooorm** [reply]
> But the whole reason I wanted to do the provider or injection approach so I do not have to Pass the components manually.

If you are going to use the default providers, you still need to define components in them.

If you are using the guide that you are, on how to inject components, you still need to inject components from your manual provider.

Providers have nothing to do with this goal you state. Components still come from somewhere.

> I just want the button automatically injected just the way Next.js does.

This question has nothing to do with injecting components. What you see is unrelated to Next. It is how JavaScript/bundlers works.

What you ask for, is how to be able to **explicitly import components from MDX**, while using `compile` and `run`.

That is of course possible. But cannot work how you think it’ll work.
